### PR TITLE
Views for Cart (add/remove)

### PR DESCRIPTION
**Pull request by NguyenTienDat377**
Write add/remove cart function with OrderSummaryView

### CODE REVIEW

File: bookstore/settings.py [hunk 1]
```diff
@@ -154,3 +154,8 @@
 ACCOUNT_EMAIL_VERIFICATION = "none"
 
 LOGIN_REDIRECT_URL = "/"
+
+INSTALLED_APPS[
```
Comment (HaruKatou) [inline bookstore/settings.py]: There exists an `INSTALLED_APPS` already, why created a new one?

File: products/views.py [hunk 2]
```diff
@@ -1,3 +1,71 @@
...
+        context = {
+            'cart_items': cartItems
+        }
+
+class PaymentView(View):
+    def get(self, *args, **kwargs):
+        cart = Cart.objects.get_object_or_404(user=self.request.user)
+        address = UserAddress.objects.get(user=self.request.user)
+        if address.
+
```
Comment (cdf144) [inline products/views.py]: Unfinished function definition

File: products/views.py [hunk 3]
```diff
@@ -1,3 +1,71 @@
...
+        cartItems = CartItems.objects.filter(cart=cart, book=book).first()
+        if cartItems.exists():
+            cartItems.quantity += 1
+            cartItems.save()
+            messages.info(request, 'Book quantity updated')
+        else:
+            cartItems.objects.create(cart=cart,book=book, quantity=1)
+            messages.info(request, 'Succesfully added book')
+    else:
+        currTime = timezone.now()
```
Comment (cdf144) [inline products/views.py]: `datetime.timezone` doesn't have member `now()`. Perhaps you meant `datetime.datetime.now()`?

File: products/forms.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    class Meta:
+        model = Book
+        fields = "__all__"
+
+        labels = {
+            'title' : 'Title',
+            'author' : 'Author',
+            'created_at' : 'Publication Date',
+            'updated_at' : 'Republication Date',
+            'publisher' : 'Publisher',
```
Comment (HaruKatou) [inline products/forms.py]: we don't store publisher anymore.

**Review by HaruKatou — DISMISSED** [review]

File: products/views.py [hunk 5]
```diff
@@ -1,3 +1,71 @@
...
+from django.contrib.auth.decorators import login_required
+from django.contrib import messages
+from django.http import HttpResponseRedirect
 
+from models import Book, CartItems, Cart, UserAddress
+from datetime import datetime, timezone
 # Create your views here.
+@login_required
+def add_to_cart(request, book, user):
+    carts = Cart.objects.get_object_or_404(user=user)
```
Comment (cdf144) [inline products/views.py:15]: `Cart.objects` doesn't exist?

**Review by cdf144 — CHANGES_REQUESTED** [review]
Lot's of problems in `views.py`

File: products/views.py [hunk 6]
```diff
@@ -1,3 +1,71 @@
...
+                cartItem.delete()
+                messages.info(request, 'Removed book from cart')
+            else:
+                cartItem.quantity -= 1
+                cartItem.save()
+                messages.info(request, 'Remove a book from already exist book')
+        except cartItem.DoesNotExist:
+            messages.info(request, 'Book not found')
+    else:
+        messages.info(request, 'You need to create your cart')
```
Comment (cdf144) [inline products/views.py]: From what I understand, this function doesn't need to be a View, so remove this.

File: products/views.py [hunk 7]
```diff
@@ -1,3 +1,71 @@
...
+    def get(self, *args, **kwargs):
+        cart = Cart.objects.get_object_or_404(user=self.request.user)
+        address = UserAddress.objects.get(user=self.request.user)
+        if address.
+
+
+
+
+
+
```
Comment (cdf144) [inline products/views.py]: Trailing lines

**Review by cdf144 — CHANGES_REQUESTED** [review]

File: products/admin.py [hunk 8]
```diff
@@ -1,4 +1,12 @@
 from django.contrib import admin
 from products.models import *
```
Comment (cdf144) [inline products/admin.py]:
```suggestion
from products.models import Book, Category
```

File: products/views.py [hunk 9]
```diff
@@ -1,3 +1,192 @@
...
+            # If page is not an integer, deliver first page
+            search_results = paginator.page(1)
+        except EmptyPage:
+            # If page is out of range (e.g. 9999), deliver last page of results
+            search_results = paginator.get_page(paginator.num_pages)
+
+    return render(request, "search.html", {"search_results": search_results})
+
 
 # Create your views here.
```
Comment (cdf144) [inline products/views.py]:
```suggestion
```

**Review by cdf144 — CHANGES_REQUESTED** [review]

File: products/forms.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+"""
+
+
+class CheckoutForm(forms.Form):
+    shipping_address = forms.CharField(max_length=255)
+    shipping_city = forms.CharField(max_length=255)
+    shipping_country = forms.CharField(max_length=255)
+    postal_code = forms.CharField(max_length=255)
+    mobile = forms.CharField(max_length=255)
+    #payment_method = forms.ChoiceField(choices=[UserPayment.PAYMENT_CHOICES])
```
Comment (cdf144) [inline products/forms.py]:
```suggestion
    payment_method = forms.ChoiceField(choices=UserPayment.PAYMENT_CHOICES)
```
Try to implement `payment_method` **in the next Pull Request**. You don't have to do this in this current Pull Request yet.

File: products/views.py [hunk 11]
```diff
@@ -1,3 +1,192 @@
...
+    def get(self, *args, **kwargs):
+        order = Order.objects.filter(user=self.request.user).first()
+        if order.address:
+            context = {
+                'order': order
+            }
+            return render(self.request, 'payment.html')
+        else:
+            messages.info('You do not have an billing address yet')
+            return redirect('checkout')
```
Comment (cdf144) [inline products/views.py:135]: Right now this is unused (and is also broken), so either delete or fix it if you use it later.

File: products/forms.py (hunk 10 above)
Comment (NguyenTienDat377) [reply products/forms.py]: Yessir

File: products/views.py (hunk 11 above)
Comment (NguyenTienDat377) [reply products/views.py:135]: I intend to use it to do Stripe and Paypal

File: products/admin.py (hunk 8 above)
Comment (NguyenTienDat377) [reply products/admin.py]: Fixed

File: templates/checkout.html [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+    <div class="form-group">
+        <label for="id_mobile">Mobile Number:</label>
+        <input type="text" name="mobile" id="id_mobile">
+    </div>
+
+    <div class="form-group">
+        <label for="id_payment_method">Payment Method:</label>
+        <select name="payment_method" id="id_payment_method">
+            {% for choice_value, choice_text in form.PAYMENT_CHOICES %}
```
Comment (cdf144) [inline templates/checkout.html]:
```suggestion
            {% for choice_value, choice_text in PAYMENT_CHOICES %}
```

File: products/views.py [hunk 13]
```diff
@@ -1,3 +1,190 @@
...
+    def post(self, *args, **kwargs):
+        form = CheckoutForm(self.request.POST or None)
+        print(self.request.POST)
+        if form.is_valid():
+            shipping_address = form.cleaned_data.get('shipping_address')
+            city = form.cleaned_data.get('shipping_city')
+            country = form.cleaned_data.get('shipping_country')
+            postal_code = form.cleaned_data.get('postal_code')
+            mobile = form.cleaned_data.get('mobile')
+            choice = form.cleaned_data.get('payment_method')
```
Comment (cdf144) [inline products/views.py]:
```suggestion
            payment_method = form.cleaned_data.get("payment_method")
            expiry_date = datetime.now()  # leave as random for now
            user_payment = UserPayment.objects.create(
                user=self.request.user,
                payment_type=payment_method,
                expiry_date=expiry_date,
            )
            user_payment.save()
```

File: products/views.py [hunk 14]
```diff
@@ -1,3 +1,190 @@
...
+from django.http import HttpResponse
+from django.template import loader
+from django.core.exceptions import ObjectDoesNotExist
+from django.shortcuts import render, get_object_or_404, redirect
+from django.views.generic import View
+from django.contrib.auth.decorators import login_required
+from django.contrib import messages
 
-# Create your views here.
+from .models import *
```
Comment (cdf144) [inline products/views.py]:
```suggestion
from .models import Book, Cart, CartItems, Order, OrderDetail, UserAddress, UserPayment
```

**Review by cdf144 — CHANGES_REQUESTED** [review]

File: products/forms.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        ("CB", "Cash on Delivery"),
+]
+
+class CheckoutForm(forms.Form):
+    shipping_address = forms.CharField(max_length=255)
+    shipping_city = forms.CharField(max_length=255)
+    shipping_country = forms.CharField(max_length=255)
+    postal_code = forms.CharField(max_length=255)
+    mobile = forms.CharField(max_length=255)
+    payment_method = forms.ChoiceField(choices=[UserPayment.PAYMENT_CHOICES])
```
Comment (cdf144) [inline products/forms.py]:
```suggestion
    payment_method = forms.ChoiceField(choices=UserPayment.PAYMENT_CHOICES)
```

**Review by cdf144 — CHANGES_REQUESTED** [review]

File: products/views.py [hunk 16]
```diff
@@ -1,3 +1,190 @@
...
+
+class CheckoutView(View):
+    def get(self, *args, **kwargs):
+        try:
+            cart = Cart.objects.get(created_by=self.request.user)
+            form = CheckoutForm()
+            context = {
+                'Form': form,
+                'Order': cart
+            }
```
Comment (cdf144) [inline products/views.py]:
```suggestion
            context = {
                'Form': form,
                'Order': cart,
                'PAYMENT_CHOICES': PAYMENT_CHOICES
            }
```

File: products/views.py [hunk 17]
```diff
@@ -1,3 +1,190 @@
...
+from django.core.exceptions import ObjectDoesNotExist
+from django.shortcuts import render, get_object_or_404, redirect
+from django.views.generic import View
+from django.contrib.auth.decorators import login_required
+from django.contrib import messages
 
-# Create your views here.
+from .models import *
+from datetime import datetime
+from .forms import CheckoutForm
```
Comment (cdf144) [inline products/views.py]:
```suggestion
from .forms import CheckoutForm, PAYMENT_CHOICES
```

**Review by cdf144 — CHANGES_REQUESTED** [review]

**Review by cdf144 — APPROVED** [review]

**Comment by cdf144** [comment]
@NguyenTienDat377
Thanks for the contribution.

**Pull request merged**